plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/AbstractRewriteCast.java

```diff
@@ -0,0 +1,100 @@
+import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.expression.Call;
+import io.trino.spi.expression.ConnectorExpression;
+import io.trino.spi.expression.Variable;
+import io.trino.spi.type.Type;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.variable;
+import static io.trino.spi.expression.StandardFunctions.CAST_FUNCTION_NAME;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractRewriteCast
+        implements ProjectFunctionRule<JdbcExpression, ParameterizedExpression>
+{
+    private static final Capture<Variable> VALUE = newCapture();
+
+    private final BiFunction<ConnectorSession, Type, String> toTargetType;
+
+    protected abstract boolean pushdownSupported(JdbcTypeHandle sourceType, Type targetType);
+
+    protected abstract Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type);
+
+    protected String buildCast(@SuppressWarnings("unused") Type sourceType, @SuppressWarnings("unused") Type targetType, String expression, String castType)
+    {
+        return "CAST(%s AS %s)".formatted(expression, castType);
+    }
+
+    public AbstractRewriteCast(BiFunction<ConnectorSession, Type, String> toTargetType)
+    {
+        this.toTargetType = requireNonNull(toTargetType, "toTargetType is null");
+    }
+
+    @Override
+    public Pattern<Call> getPattern()
+    {
+        return call()
+                .with(functionName().equalTo(CAST_FUNCTION_NAME))
+                .with(argumentCount().equalTo(1))
+                .with(argument(0).matching(variable().capturedAs(VALUE)));
+    }
+
+    @Override
+    public Optional<JdbcExpression> rewrite(ConnectorExpression castExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Variable variable = captures.get(VALUE);
+        JdbcTypeHandle sourceTypeJdbcHandle = ((JdbcColumnHandle) context.getAssignment(variable.getName())).getJdbcTypeHandle();
+        Type targetType = castExpression.getType();
+
+        if (!pushdownSupported(sourceTypeJdbcHandle, targetType)) {
+            return Optional.empty();
+        }
+
+        Optional<ParameterizedExpression> value = context.rewriteExpression(variable);
+        if (value.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<JdbcTypeHandle> targetTypeJdbcHandle = toJdbcTypeHandle(targetType);
+        if (targetTypeJdbcHandle.isEmpty()) {
+            return Optional.empty();
+        }
```

Comment:
Can we have it before rewriting the expression so that we could skip rewriting if the target type is not supported.